### DISCUSSION TITLE
How does mitmproxy track HTTP 1.1 requests and responses?

### DISCUSSION THREAD

**Discussion by ghost** [mitmproxy/mitmproxy#5914] [Q&A]
In HTTP/2 the protocol has the concept of a stream ID. So one "conversation" (request/response) is always marked with the same stream ID. This means it's easy to parse the frames into logical request/response conversations.

But how can we do this for HTTP 1.1? How does mitmproxy do it? In mitmproxy vocabulary, how do you connect `flow.request` with `flow.response`?

Does anyone have references to the parts of the codebase where this is handled?

**Comment by mhils** [comment]
Generally speaking we use the same concept of streams for all HTTP versions, for HTTP/1 we just make up stream IDs. Our HTTP/1 logic is in https://github.com/mitmproxy/mitmproxy/blob/main/mitmproxy/proxy/layers/http/_http1.py if you want to take a look. The illustration in https://mitmproxy.org/posts/new-proxy-core/ may also help understand things. :)

**Reply by ghost** [reply]
Thanks that helped a lot!
